File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def split_part(file):
    head = ""
    number = ""
    tail = ""
    for i in range(len(file)):
        if file[i].isdigit():
            head = file[:i]
            for j in range(i, len(file)):
                if not file[j].isdigit():
                    number = file[i:j]
                    tail = file[j:]
                    break
            else:
                number = file[i:]
            break

    return head, number, tail


def solution(files):
    answer = []
    file_dict = {}
    for i in files:
        head, number, tail = split_part(i)
        file_dict[i] = (head, number, tail)

    sorted_files = sorted(
        file_dict.items(), key=lambda x: (x[1][0].lower(), int(x[1][1]))
    )

    for i in sorted_files:
        answer.append(i[0])
    return answer

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.java
import java.util.*;
class fileName {
    String head;
    int num;
    String tail;
    String file_name;
    int idx;
    fileName(String file, int idx){
        this.idx = idx;
        this.file_name = file;
        int i = 0;
        while(!Character.isDigit(file.charAt(i))){
            i++;
        }
        this.head = file.substring(0,i);
        int j = i;
        while(j < file.length() && Character.isDigit(file.charAt(j))){
            j++;
        }
        this.num = Integer.parseInt(file.substring(i,j));
        this.tail = file.substring(j);
    }
}
class Solution {
    public String[] solution(String[] files) {
        List<fileName> list = new ArrayList<>();
        for(int i = 0; i<files.length; i++){
            list.add(new fileName(files[i], i));
        }
        list.sort(
                Comparator.comparing((fileName a)->a.head.toLowerCase())
                        .thenComparingInt(a->a.num)
        );
        String[] answer = new String[list.size()];
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i).file_name;
        }
        return answer;
    }
}